File: source/central_unit/communication.py
# Copyright 2016 Noak Ringman, Emil Segerbäck, Robin Sliwa, Frans Skarman, Hannes Tuhkala, Malcolm Wigren, Olav Övrebö
# 

# This file is part of LiTHe Hex.

# LiTHe Hex is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# LiTHe Hex is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with LiTHe Hex.  If not, see <http://www.gnu.org/licenses/>.

import spidev
import time
import pdb

MOTOR_ADDR = (0, 0)
SENSOR_ADDR = (0, 1)

DATA_REQ    = 0x02

SEND_FAIL   = 0x1F
ACK         = 0x0F

MAX_BYTE_SIZE = 255
MAX_16BIT_SIZE = MAX_BYTE_SIZE**2

SET_OBSTACLE = 0x03

SET_SERVO_SPEED = 0x20
SET_SERVO_SPEED_LENGTH = 2

WALK = 0x21
WALK_LENGTH = 3

RETURN_TO_NEUTRAL = 0x05

SERVO_STATUS = 0x22
MOTOR_DEBUG = 0x23

SENSOR_DATA = 0x24

SENSOR_CORRIDOR = 0x25


class InvalidCommandException(Exception):

    def __init__(self, message, command=""):
        self.message = message
        self.command = command

    def __repr__(self):
        return self.message

    def __str__(self):
        return self.message


class CommunicationError(Exception):

    def __init__(self, message, cause=None):
        self.message = message
        self.cause = cause

    def __repr__(self):
        return self.message

    def __str__(self):
        return self.message


def _calculate_parity(data):
    data_str = "{0:032b}".format(data)
    num_ones = data_str.count("1")
    if num_ones % 2 == 0:
        return 0
    else:
        return 1


def _add_parity(type_, msg):
    msg_parity = _calculate_parity(msg)
    result = type_ << 2
    result |= msg_parity << 1
    type_parity = _calculate_parity(result)
    result |= type_parity
    return result
    

def _send_bytes(spi, *data):
    # check if all are bytes
    if sum([(not isinstance(x, int)) or x > MAX_BYTE_SIZE or x < 0 for x in data]):
        raise InvalidCommandException("Data sequence {} contains non-bytes".format(data))
    spi.writebytes([66])
    spi.writebytes(list(data))
    return spi.readbytes(1)
        

def _recieve_muliple_bytes(spi, type_):
    length = spi.readbytes(1)[0]
    return spi.readbytes(length)
    

def _recieve_single_byte(spi, type_):
    return spi.readbytes(1)[0]


def _recieve_bytes(spi):
    type_ = spi.readbytes(1)[0]

    # if the most significant bit in the
    # type message is 1, then the message
    # is of variable length
    if (type_ & 0x80):
        return _recieve_muliple_bytes(spi, type_)
    else:
        return _recieve_single_byte(spi, type_)


def _is_multibyte_msg(data_id):
    tmp = data_id << 2
    tmp_str = '{0:08b}'.format(tmp)
    if tmp[0] == '1':
        return True
    return False


def _request_data(spi, data_id):
    # response = spi.xfer2([DATA_REQ, data_id])
    # pdb.set_trace()
    response = _send_bytes(spi, _add_parity(DATA_REQ, data_id), data_id)
    _check_response(response)
    return _recieve_bytes(spi)


def _select_device(spi, addr):
    spi.open(*addr)


def _check_response(response):
    if isinstance(response, list):
        _check_response(response[0])
    elif response == SEND_FAIL:
        raise CommunicationError("Send fail received")
    elif response != ACK:
        raise CommunicationError("No acknowledge message received")


def _get_total_msg(*data):
    length = len(data)
    res = str(length) if length > 1 else ""
    for d in data:
        res += str(d)
    return int(res)


def communication_init():
    spi = spidev.SpiDev()
    spi.open(*MOTOR_ADDR)
    spi.max_speed_hz = 25000
    return spi


def set_obstacle_mode(spi, value):
    if value not in (True, False):
        raise InvalidCommandException("Value \"{}\" is not a valid mode.".format(value))
    # yes i'm paranoid
    value = 0x01 if value else 0x00
    _select_device(spi, MOTOR_ADDR)
    response = _send_bytes(spi, _add_parity(SET_OBSTACLE, value), value)
    _check_response(response)


def set_servo_speed(spi, speed):
    if speed < 0 or speed > MAX_16BIT_SIZE:
        raise InvalidCommandException("Speed \"{}\" is not a 16-bit value".format(speed))
    _select_device(spi, MOTOR_ADDR)
    least = speed & 0x00FF
    most = (speed & 0xFF00) >> 8
    total_msg = _get_total_msg(SET_SERVO_SPEED_LENGTH, least, most)
    response = _send_bytes(spi, _add_parity(SET_SERVO_SPEED, total_msg), 
                           SET_SERVO_SPEED_LENGTH, least, most)
    _check_response(response)


def walk(spi, x_speed, y_speed, turn_speed):
    _select_device(spi, MOTOR_ADDR)
    total_msg = _get_total_msg(x_speed, y_speed, turn_speed)
    response = _send_bytes(spi, _add_parity(WALK, total_msg),
                           WALK_LENGTH, x_speed, y_speed, turn_speed)
    _check_response(response)


def back_to_neutral(spi):
    _select_device(spi, MOTOR_ADDR)
    pass


def get_servo_status(spi):
    _select_device(spi, MOTOR_ADDR)
    return _request_data(spi, SERVO_STATUS)


def get_motor_debug(spi):
    _select_device(spi, MOTOR_ADDR)
    return _request_data(spi, MOTOR_DEBUG)


def get_sensor_data(spi):
    _select_device(spi, SENSOR_ADDR)
    return _request_data(spi, SENSOR_DATA)


def get_corridor_data(spi):
    _select_device(spi, SENSOR_ADDR)
    return _request_data(spi, SENSOR_DATA)

File: source/communication/communication.c
// Copyright 2016 Noak Ringman, Emil Segerbäck, Robin Sliwa, Frans Skarman, Hannes Tuhkala, Malcolm Wigren, Olav Övrebö

// This file is part of LiTHe Hex.

// LiTHe Hex is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// LiTHe Hex is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with LiTHe Hex.  If not, see <http://www.gnu.org/licenses/>.

#include <stdbool.h>
#include "spi.h"
#include "communication.h"
#include "../sensor/table.h"
#include <stdio.h>

uint8_t replay_msg_id[5] = {0x02, 0x03, 0x04, 0x20, 0x05};

bool check_partiy();

void send_replay_sensor(uint8_t current_msg);

void send_replay_motor(uint8_t current_msg);

bool message_require_replay(uint8_t current_msg);

uint8_t get_current_id(Frame* frame_recv);

void get_new_frame(Frame* frame_recv);

void send_frame(Frame* frame_send);

void on_spi_recv() {
	Frame frame_recv;
	get_new_frame(&frame_recv);
	uint8_t current_id = get_current_id(&frame_recv);
	
	bool success = check_partiy(&frame_recv); 
	if(success) { // continue if message ok
		spi_transmit_ack();

		bool replay = message_require_replay(current_id);
		if(replay) { // send a replay to central-unit
			#ifdef MainTable
				send_replay_sensor(current_id);
			#endif
			//#ifdef //TODO send replay for motor if some file is def
				//send_replay_motor();
		} else { // 
			//#ifdef //TODO send replay for motor if some file is def
			//control_motor();
		}
	} else { // Something was wrong with message
		spi_transmit_byte(0x1F);
	}
}

bool check_partiy(Frame* frame) {
	// check parity for control_byte
	uint8_t byte = frame->control_byte;
	bool parity_con = false;
	while(byte) {
		parity_con = !parity_con;
		byte &= (byte - 1);
	}

	// check parity for length and message bytes
	bool parity_msg = false;
	for(uint8_t i = 0; i < frame->len; i++) {
		uint8_t msg = frame->msg[i];
		while(msg) {
			parity_con = !parity_con;
			msg &= (msg - 1);
		}
	}
	
	return (parity_con == (frame->control_byte & 0x01)) && (parity_msg == (frame->control_byte & 0x02));
}

void calculate_parity(Frame* frame) {
	// calculate parity for length and message bytes
	bool parity_msg = false;
	for(uint8_t i = 0; i < frame->len; i++) {
		uint8_t msg = frame->msg[i];
		while(msg) {
			parity_msg = !parity_msg;
			msg &= (msg - 1);
		}
	}

	// Set parity bit for len and message
	if(parity_msg) {
		frame->control_byte | 0x02;
	}
	
	// calculate parity for control_byte
	uint8_t byte = frame->control_byte;
	bool parity_con = false;
	while(byte) {
		parity_con = !parity_con;
		byte &= (byte - 1);
	}

	// Set parity bit for control_byte
	if(parity_con) {
		frame->control_byte | 0x01;
	}

}

void get_new_frame(Frame* frame_recv) {
	frame_recv->control_byte = spi_recieve_byte();
	if(frame_recv->control_byte & 0x80) { // msg is one byte long
		frame_recv->len = 0;
		frame_recv->msg[0] = spi_recive_byte();
	} else { // msg is more than one byte long
		frame_recv->len = spi_recieve_byte();
		for(uint8_t i = 0; i < frame_recv->len; i++) {
			frame_recv->msg[i] = spi_recieve_byte();
		}
	}
}

uint8_t get_current_id(Frame* frame_recv) {
	return (frame_recv->control_byte & 0xFC) >> 2;
}

bool message_require_replay(uint8_t current_msg) {
	for(uint8_t i = 0; i < 7; i++) {
		if(current_msg == replay_msg_id[i]) {
			return true;
		}
	}
}

void send_replay_sensor(uint8_t current_id) {
	Frame frame_send_1;
	Frame frame_send_2;
	if(current_id == 0x02) {
		frame_send_1.control_byte = 0x20 << 2;
		send_sensor_data(&frame_send_1);
		frame_send_2.control_byte = 0x21 << 2;
		send_sensor_wall_data(&frame_send_2);
	}
	send_frame(&frame_send_1);
	send_frame(&frame_send_2);
}

void send_frame(Frame* frame) {
	calculate_parity(frame);
	spi_transmit_byte(frame->control_byte);
	spi_transmit_byte(frame->len);
	for(uint8_t i = 0; i < frame->len; i++) {
		spi_transmit_byte(frame->msg[i]);
	}
}

void control_motor(uint8_t current_msg) {
	switch(current_msg){
		case 0x03 :
			// Toggle obstacle
			break;
		case 0x04 :
			// Set speed
			break;
		case 0x20 :
			// Go command 
			break;
		case 0x05 :
			// Return to neutral
			break;
	}
}

void send_replay_motor(uint8_t current_msg) {
	Frame frame_send_status;
	Frame frame_send_string;
	Frame frame_send_obstacle;
	frame_send_status.control_byte = 0x20 << 2;
	// Send servo status
	frame_send_string.control_byte = 0x21 << 2;
	// Send debug string
	frame_send_obstacle.control_byte = 0x03 << 2;
	// Hinder here?
	send_frame(&frame_send_status);
	send_frame(&frame_send_string);
	send_frame(&frame_send_obstacle);
}